File: backend/src/models/Telemetry.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface ITelemetry extends Document {
  driverId: string;
  orderId: string | null;
  location: {
    lat: number;
    lng: number;
  };
  speed: number; // km/h
  heading: number; // degrees
  timestamp: Date;
}

const TelemetrySchema: Schema = new Schema({
  driverId: { type: String, required: true },
  orderId: { type: String, default: null },
  location: {
    lat: { type: Number, required: true },
    lng: { type: Number, required: true }
  },
  speed: { type: Number, required: true },
  heading: { type: Number, required: true },
  timestamp: { type: Date, default: Date.now }
});

// Index for query speed on historical data
TelemetrySchema.index({ driverId: 1, timestamp: -1 });

export default mongoose.model<ITelemetry>('Telemetry', TelemetrySchema);

File: backend-java/src/main/java/com/fleet/dispatch/model/Telemetry.java
package com.fleet.dispatch.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.Date;

@Document(collection = "telemetry")
public class Telemetry {
    @Id
    private String id;
    private String driverId;
    private String orderId;
    private Location location;
    private double speed;
    private double heading;
    private Date timestamp = new Date();

    public Telemetry() {}

    public Telemetry(String driverId, String orderId, Location location, double speed, double heading) {
        this.driverId = driverId;
        this.orderId = orderId;
        this.location = location;
        this.speed = speed;
        this.heading = heading;
        this.timestamp = new Date();
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getDriverId() {
        return driverId;
    }

    public void setDriverId(String driverId) {
        this.driverId = driverId;
    }

    public String getOrderId() {
        return orderId;
    }

    public void setOrderId(String orderId) {
        this.orderId = orderId;
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    public double getSpeed() {
        return speed;
    }

    public void setSpeed(double speed) {
        this.speed = speed;
    }

    public double getHeading() {
        return heading;
    }

    public void setHeading(double heading) {
        this.heading = heading;
    }

    public Date getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }
}
